### PR TITLE
allow support orm using routing orm bundle

### PR DESCRIPTION
**Pull request by almacbe**
| Q             | A
| ------------- | ---
| Branch?       | "master" 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| License       | MIT

Some ideas I was working to try to support orm using a new bundle. You can follow the previous conversation #223 

Let's discuss it and I could continue after the feedback that you give me.

### CODE REVIEW

**Comment by almacbe** [comment]
thanks, I am going to work on this tomorrow.

If you could, have a look the [repo of routing auto orm](https://github.com/wearemarketing/routing-auto-orm-bundle) as well.

File: src/Adapter/AutoRouteRefreshCommandAdapterInterface.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * @author WAM Team <develop@wearemarketing.com>
+ */
+interface AutoRouteRefreshCommandAdapterInterface
+{
+    /**
+     * Return all content of given class. Those classes will refresh their auto routes.
+     *
+     * @return array
```
[the remark this answers is not in the document]
Comment (almacbe) [reply src/Adapter/AutoRouteRefreshCommandAdapterInterface.php]: An autoroute has to be always an object. So, let's use object[]

**Comment by almacbe** [comment]
@dantleech I have changed the name of the interface.

@dbu I have done the changes that you suggest.

File: .travis.yml [hunk 2]
```diff
@@ -51,6 +51,7 @@ matrix:
 
 before_install:
   - phpenv config-rm xdebug.ini || true
+  - phpenv config-add travis.php.ini
```
Comment (dantleech) [inline .travis.yml]: What does this do out of curiosity?

File: src/Adapter/AutoRouteRefreshAdapterInterface.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    public function getAllContent(string $classFqn);
+
+    /**
+     * Return the identifier of an auto route.
+     *
+     * @param mixed $autoRouteableContent
+     *
+     * @return mixed
+     */
+    public function getIdentifier($autoRouteableContent);
```
Comment (dantleech) [inline src/Adapter/AutoRouteRefreshAdapterInterface.php]: These are two methods could be resused indepently of one another, if we are segregating interfaces then these should be two seperate interfaces, otherwise we could just put them all in the `Adapter` interface. Not having worked on this for years, I would be tempted to just add them to the `AdapterInterface`?

We should return an [iterable](http://php.net/manual/en/language.types.iterable.php) type from `getAllContent` - it is very dangerous to perform f.e. an ODM query on 10,000 entities and hydrate them all as an array.

File: src/Adapter/PhpcrOdmAdapter.php [hunk 4]
```diff
@@ -243,6 +243,24 @@ public function findRouteForUri($uri, UriContext $uriContext)
...
+    {
+        $qb = $this->dm->createQueryBuilder();
+        $qb->from()->document($classFqn, 'a');
+        $q = $qb->getQuery();
+        $content = $q->getResult();
+
+        return $content;
+    }
+
+    public function getIdentifier($autoRouteableContent)
```
Comment (dantleech) [inline src/Adapter/PhpcrOdmAdapter.php]: Missing `@inheritdoc` from these methods.

File: src/Command/RefreshCommand.php [hunk 5]
```diff
@@ -51,54 +99,54 @@ public function configure()
         $this->addOption('session', null, InputOption::VALUE_OPTIONAL, 'The session to use for this command');
     }
 
+    protected function interact(InputInterface $input, OutputInterface $output)
+    {
+        parent::interact($input, $output);
+
+        // TODO: check if it is phpcr check for session or if it is orm do not accept session
+    }
```
Comment (dantleech) [inline src/Command/RefreshCommand.php]: Is this necessary? Can we remove it? If this a foundation for another feature, then it can be another PR.

File: src/Command/RefreshCommand.php [hunk 6]
```diff
@@ -51,54 +99,54 @@ public function configure()
...
-        DoctrineCommandHelper::setApplicationPHPCRSession(
-            $this->getApplication(),
-            $session
-        );
+        if ('doctrine_phpcr_odm' === $this->adapterName) {
+            DoctrineCommandHelper::setApplicationPHPCRSession(
+                $this->getApplication(),
+                $session
+            );
+        }
```
Comment (dantleech) [inline src/Command/RefreshCommand.php]: Code smell. This could _maybe_ be part of a PHPCR-ODM refresh routes service as mentioned below. I actually forget why we need this at all, seems like a legacy mistake.

File: src/DependencyInjection/CmfRoutingAutoExtension.php [hunk 7]
```diff
@@ -65,6 +67,14 @@ public function load(array $configs, ContainerBuilder $container)
                 $adapterName = 'doctrine_phpcr_odm';
             }
             $container->setParameter('cmf_routing_auto.persistence.phpcr.route_basepath', $config['persistence']['phpcr']['route_basepath']);
+
+            $commandService = $container->getDefinition('cmf_routing_auto.auto_route_refresh_command');
+            $commandService->replaceArgument(2, new Reference('doctrine_phpcr.odm.default_document_manager'));
+            $commandService->replaceArgument(3, new Reference('cmf_routing_auto.adapter.phpcr_odm'));
+        }
```
Comment (dantleech) [inline src/DependencyInjection/CmfRoutingAutoExtension.php]: It might make more sense to have a `RefreshRoutes` service which is passed to the command. This service could then be invoked from other contexts (e.g. a web UI). You could then create an implementation for either the ORM or the PHPCR-ODM, and you will not be limited to the dependencies accepted by the command.

For logging you could pass a closure.

File: .travis.yml (hunk 2 above)
Comment (dantleech) [reply .travis.yml]: Ah I see -- memory limit :scream: I guess this is due to the composer setup?

File: tests/Functional/Command/RefreshCommandTest.php [hunk 8]
```diff
@@ -47,13 +47,19 @@ public function testCommand()
...
+        $adapter = $container->get('cmf_routing_auto.adapter.phpcr_odm');
+
         $application = $this->getApplication();
         $input = new ArrayInput([
             'foo:bar',
         ]);
         $output = new NullOutput();
-        //$output = new StreamOutput(fopen('php://stdout', 'w'));
-        $command = new RefreshCommand();
+//        $output = new StreamOutput(fopen('php://stdout', 'w'));
```
Comment (dantleech) [inline tests/Functional/Command/RefreshCommandTest.php]: If it was commented, then we may as well remove it.

**Comment by dantleech** [comment]
Actually, the refresh command is less than ideal in it's current form. I wonder if it would make more sense to just make it PHPCR-ODM only (only register it if that adapter is used) and have a separate command in the ORM bundle?

**Comment by almacbe** [comment]
I am working on this and I have made some changes. The idea behind is don't load the phpcr services and command and allow the routing orm bundle load the new services and command to support orm. With this idea, we can keep the same features than now. what do you think @dantleech ?

There is an error on travis related to a class that not exists and I have research about a solution and it seems we have to upgrade phpunit. I have seen that I have to modify travis file to do it but it is auto generated. I have checked that it is a issue on other repos, so, could you give some help about this? @dbu

File: src/CmfRoutingAutoBundle.php [hunk 9]
```diff
@@ -27,6 +28,11 @@ public function build(ContainerBuilder $container)
         $this->buildPhpcrCompilerPass($container);
     }
 
+    public function registerCommands(Application $application)
+    {
+        // Do not allow discover automatically commands
```
Comment (dantleech) [inline src/CmfRoutingAutoBundle.php]: "disable automatic discovery of commands" ?

File: src/CmfRoutingAutoBundle.php [hunk 10]
```diff
@@ -50,7 +56,7 @@ private function buildPhpcrCompilerPass(ContainerBuilder $container)
                         realpath(__DIR__.'/Resources/config/doctrine-model') => 'Symfony\Cmf\Bundle\RoutingAutoBundle\Model',
                     ],
                     ['cmf_routing_auto.persistence.phpcr.manager_name'],
-                    false,
+                    'cmf_routing.backend_type_phpcr',
```
Comment (dantleech) [inline src/CmfRoutingAutoBundle.php]: what is the purpose of this change?

**Comment by ElectricMaxxx** [comment]
Hi, thanks a lot for the work. If both won't do it before i will go over it tomorrow night. Just a little hint for routing ORM support: there are two PR open in routing-bundle to verrify ORM functionality and to make its usage ckearer. Just if you'd like to see ORM working.

File: composer.json [hunk 11]
```diff
@@ -28,7 +28,8 @@
     "suggest": {
         "doctrine/phpcr-odm": "To enable support for the PHPCR ODM documents",
         "doctrine/phpcr-bundle": "To enable support for the PHPCR ODM documents",
-        "symfony/yaml": "To be able to load Yaml configuration files (^2.1)"
+        "symfony/yaml": "To be able to load Yaml configuration files (^2.1)",
+        "symfony-cmf/routing-auto-orm-bundle": "To enable support for the Doctrine ORM Entities"
```
Comment (dantleech) [inline composer.json]: this package doesn't exist yet?

File: src/DependencyInjection/CmfRoutingAutoExtension.php [hunk 12]
```diff
@@ -65,6 +65,12 @@ public function load(array $configs, ContainerBuilder $container)
                 $adapterName = 'doctrine_phpcr_odm';
             }
             $container->setParameter('cmf_routing_auto.persistence.phpcr.route_basepath', $config['persistence']['phpcr']['route_basepath']);
+
+            $loader->load('commands.xml');
+        }
+
+        if (isset($bundles['CmfRoutingAutoOrmBundle'])) {
+            $hasProvider = true;
```
Comment (dantleech) [inline src/DependencyInjection/CmfRoutingAutoExtension.php]: Is it not enough to set the adapter name to `doctrine_orm` in the config?

File: src/CmfRoutingAutoBundle.php (hunk 9 above)
Comment (almacbe) [reply src/CmfRoutingAutoBundle.php]: Thanks.

File: src/CmfRoutingAutoBundle.php (hunk 10 above)
Comment (almacbe) [reply src/CmfRoutingAutoBundle.php]: it allows to disable the mapping of phpcr documents when you load auto routing orm bundle. You can see here https://github.com/wearemarketing/routing-auto-orm-bundle/blob/master/src/CmfRoutingAutoOrmBundle.php#L26

File: composer.json (hunk 11 above)
Comment (almacbe) [reply composer.json]: It is not yet. I am going to remove it and we will add it when we created it

File: src/DependencyInjection/CmfRoutingAutoExtension.php (hunk 12 above)
Comment (almacbe) [reply src/DependencyInjection/CmfRoutingAutoExtension.php]: It isn't. We have to add more config. You can see here https://github.com/wearemarketing/routing-auto-orm-bundle/blob/master/src/DependencyInjection/CmfRoutingAutoOrmExtension.php#L47

IMHO I leave all those configs to the new bundle. What do you think?

**Comment by almacbe** [comment]
@ElectricMaxxx your welcome. We are using this at least 5 years in production, so we are so excited to share it.

Are you talking about the PR on this repo? One of them is mine and the other one inspire me how to improve what we did. I have closed mine, and now the main thread it is this.

File: src/DependencyInjection/CmfRoutingAutoExtension.php (hunk 12 above)
Comment (dantleech) [reply src/DependencyInjection/CmfRoutingAutoExtension.php]: If we are prepending the config in the ORM bundle, then shouldn't the adapter be configured here, and the exception should not be thrown?

File: src/Resources/config/phpcr-odm.xml [hunk 13]
```diff
@@ -5,7 +5,7 @@
 
     <services>
         <!-- Adapters -->
-        <service id="cmf_routing_auto.adapter.phpcr_odm" class="Symfony\Cmf\Bundle\RoutingAutoBundle\Adapter\PhpcrOdmAdapter">
+        <service id="cmf_routing_auto.adapter.phpcr_odm" class="Symfony\Cmf\Bundle\RoutingAutoBundle\Adapter\PhpcrOdmAdapter" public="true">
```
Comment (dantleech) [inline src/Resources/config/phpcr-odm.xml]: why does this service now need to be public? (I guess it's not used by the ORM bundle?)

File: src/CmfRoutingAutoBundle.php (hunk 10 above)
Comment (dantleech) [reply src/CmfRoutingAutoBundle.php]: can we not instead only add the PHPCR mappings if the PHPCR adapter is used?

Comment (dantleech) [reply src/CmfRoutingAutoBundle.php]: but in anycase, this looks like a good change cc @dbu

Comment (almacbe) [reply src/CmfRoutingAutoBundle.php]: Actually, it is a good change as well, I think it is so common to add that parameter to allow or not the mapping.

File: src/Resources/config/phpcr-odm.xml (hunk 13 above)
Comment (almacbe) [reply src/Resources/config/phpcr-odm.xml]: You are right. I've forgotten to remove this.

File: src/DependencyInjection/CmfRoutingAutoExtension.php (hunk 12 above)
Comment (almacbe) [reply src/DependencyInjection/CmfRoutingAutoExtension.php]: I am going to check it... because it makes sense what you say...

**Comment by dantleech** [comment]
Looks like some PHPUnit hell with the build... have you seen this before @ElectricMaxxx ?

File: src/CmfRoutingAutoBundle.php (hunk 10 above)
Comment (dbu) [reply src/CmfRoutingAutoBundle.php]: yep, looks sensible to me.

File: src/DependencyInjection/CmfRoutingAutoExtension.php [hunk 14]
```diff
@@ -65,6 +65,8 @@ public function load(array $configs, ContainerBuilder $container)
                 $adapterName = 'doctrine_phpcr_odm';
             }
             $container->setParameter('cmf_routing_auto.persistence.phpcr.route_basepath', $config['persistence']['phpcr']['route_basepath']);
+
+            $loader->load('commands.xml');
```
Comment (dbu) [inline src/DependencyInjection/CmfRoutingAutoExtension.php]: can we name this file phpcr_commands.xml to make things even more explicit?

**Comment by dbu** [comment]
regarding making orm auto-routing a cmf organization bundle: i would be happy to have it here if you would take the responsibility to maintain it @almacbe. that would mean i add you to the github organization and make you owner of that repository. i can also give you commit rights on the routing-auto-bundle repository.

**Comment by dbu** [comment]
fixed the build in #230, please rebase this on master.

**Comment by almacbe** [comment]
ping @dantleech @dbu

**Comment by dantleech** [comment]
Fine with me :+1:

**Comment by dbu** [comment]
thanks!

regarding the orm auto-routing bundle: we have no written rules. we follow the symfony guidelines on bundles and codestyle, and we want to be respectful to people we interact with, and take security warnings seriously. i think thats about it.

ping me when you are ready and tell me the username of your collegue, then i can create the repository so that you can push the code to it.
